### DISCUSSION TITLE
On the code sample question

### DISCUSSION THREAD

**Discussion by hongshixia** [jcreinhold/intensity-normalization#73] [Q&A] [answered]
Hello, I'm sorry for the delay. I was a medical student and a complete novice at coding. I believe that all the developers have prepared the instructions with great care, but I still can't understand them. I can only follow the API of the example to complete the method about fcm, but I still can't understand the other five methods. So could you please give me some examples about the API of other methods? Another problem is that I do not understand the use of Cli, I asked other computer friends they do not know how to use it. So can you come up with some more detailed tutorials, preferably with picture examples. Thanks a lot!

**Comment by jcreinhold** [comment, marked as answer]
I'm sorry but without more detail about where you're getting stuck, I can't help. I took a good amount of time trying to create the documentation along with examples and tutorials, so please try re-reading them carefully.

It seems that people are able to successfully use this software in their research, so without details on where exactly you're getting stuck or lost, I don't have the time to help.

If you're not familiar with Python or working in the terminal, you should first look at Python and bash tutorials, respectively.

**Reply by hongshixia** [reply]
I'm glad to hear from you, Sir. I may not have articulated my question clearly. I have read the document you wrote carefully. I've been working on it for two months.In the document you wrote, I followed the example of the fcm method for successful image processing.I did the following

```python
import nibabel as nib
from intensity_normalization.typing import Modality, TissueType
from intensity_normalization.normalize.fcm import FCMNormalize

image = nib.load("test_t1w_image.nii").get_fdata()  # assume skull-stripped otherwise load mask too

fcm_norm = FCMNormalize(tissue_type=TissueType.WM)
normalized = fcm_norm(image)

# now normalize the co-registered, corresponding T2-w image
t2w_image = nib.load("test_t2w_image.nii").get_fdata()
t2w_normalized = fcm_norm(t2w_image, modality=Modality.T2)

# to use a brain mask instead of a skull-stripped image do this:
mask = nib.load("brain_mask.nii").get_fdata()
normalized_t1w = fcm_norm(image, mask)
# the WM mask is an attribute in the class, so normalize the t2 with:
normalized_t2w = fcm_norm(t2w_image, modality=Modality.T2)

# make a new instance of the normalizer to normalize a new image, i.e.:
new_image = nib.load("test_t1w_image_2.nii")
fcm_norm = FCMNormalize(tissue_type=TissueType.WM)
normalized = fcm_norm(new_image.get_fdata())

# you can save the normalized image with nibabel as follows:
nib.Nifti1Image(normalized, new_image.affine).to_filename("normalized.nii")
```

It seems that I can only use the above form to deal with it. So I would like to ask if there are other methods (such as nyul, Z-score, etc.) in this form. I know my question may seem a little low level, but it's really the problem I'm having right now. So I want to ask. I don't know if I made it clear this time.

**Reply by hongshixia** [reply]
Excuse me, Sir. What I just wanted to say is that I can't understand your document at my level. I can only read the examples in example usage.

**Reply by jcreinhold** [reply]
[Here is an example of the Nyul normalization scheme](https://intensity-normalization.readthedocs.io/en/latest/usage.html#saving-fit-information-for-sample-based-methods).
 
[Here is an example of Z-score normalization](https://github.com/jcreinhold/intensity-normalization/issues/68).

I hope these help. Unfortunately, I don't have time to generally revamp the documentation without actionable feedback.

**Reply by hongshixia** [reply]
These are enough,Thank you,sir!

**Reply by hongshixia** [reply]
Hello, Sir. Last time I read your reply, I studied it carefully and benefited a lot. I have encountered some problems in the study process and would like to ask your advice.

1. For Nyul's method, he uses the training set to learn a standard scale, so how many individual brain images are generally needed to train? For example, if I have the data of 300 people, how many people would you recommend to train this standard scale?

2. I think this question may seem silly, but I want to make sure. Before this, I ran the code you gave me, and I found that the code did not transform the image of the learning set. Therefore, on the premise of the previous question, for example, if I have 300 data and I take out A data to train the standard scale, in the second transformation stage, can the previous A data also enter the transformation stage for transformation?

**Reply by jcreinhold** [reply]
1. I'd pick a random subset of 20-50 of the subjects to train the Nyul method on. I guess as large of a subset as your computer can handle without running out of memory.

2. If you ran:
```python
nyul_normalizer.fit(images)
normalized = [nyul_normalizer(image) for image in images]
```

and the images in `normalized` aren't normalized, then there is bug. Please raise a bug issue and fill out the relevant form. 

If you're simply asking if you can normalize the training data with the Nyul normalizer fit on the training data, the answer is yes.

**Reply by hongshixia** [reply]
Hello,  thank you for your reply.  The way I judge whether there is no normalization is to extract the radiomics features and  compare them with the previous ones.  My understanding is that this code has two stages,  one is the learning stage and the other is the conversion stage.  The general example is this.  At that time,  in order to quickly learn how to run the code, I simply used image 1,2 for the learning stage, and used image3 for the transformation stage.  It was roughly like this,  if the position of image3 before running was "D:/deskbook/study/image3", After running,  the image features of this position are different from before,  but 1 and 2 are still the same as before.  If I use image 1 and 2 as the learning phase and image 1,2,3 as the conversion  phase, after running, the three images will change in their original positions as described above,  and the image features will also change.
That's my process and the phenomenon that happens.  Do you think there is any  error in my operation?

**Reply by jcreinhold** [reply]
How are you opening and saving the images? You need to save the normalized images with the same affine matrix as the original image. If you're using nibabel to open and save the images, you can see an example of how to save the images in the [tutorial](https://intensity-normalization.readthedocs.io/en/latest/usage.html#python-api-for-normalization-methods).

Refer to the documentation of whatever package you're using to open and save images to see what information you should be passing along to correctly save the images so this shift doesn't occur.

If you're saving the images with the correct affine matrix and you're still encountering a shift in coordinates and the only potential source of the error is with `intensity-normalization`, please provide screenshots of the difference and a more thorough description of what you did, along with all of the other requested details in a bug report, and file the issue.

**Comment by hongshixia** [comment]
Yes, Sir. Thank you for your reply. I'll study hard again.

**Comment by hongshixia** [comment]
Hello, Sir. I think I may not have expressed my question yesterday
My code is as follows
![image](https://user-images.githubusercontent.com/118377176/206087198-a6b02e01-213a-4956-bf67-b9abcb1135cc.png)
At the end of the run, my result is
![image](https://user-images.githubusercontent.com/118377176/206087368-6c9708ea-de84-4453-bd88-c999d8e245c1.png)
The two images here are unchanged.


Then,
![image](https://user-images.githubusercontent.com/118377176/206087561-baf52dd5-d93c-4683-9bbf-438953397cb4.png)
Here the image is changed in its original position.

This is the process and result I wanted to express yesterday, so I want to ask you if there is any problem with this process and result.
Please excuse my intrusion.

**Reply by jcreinhold** [reply]
Sorry, I don't understand what you mean by "changed in its original position", and I'm a bit lost as to what your question is exactly. I believe you're getting stuck on some basic programming concepts and perhaps poor computer-generated translations of the documentation. Unfortunately, I can't help with either of these issues.

If you saved the image without the affine matrix, the coordinates of the normalized image will be different. Please refer to the documentation of nibabel for reference (e.g., [here](https://nipy.org/nibabel/coordinate_systems.html)).

`images` is a list of the unaltered 3D or 2D arrays (depending on the dimension of `image1.nii` and `image2.nii`). They should be unchanged when this list is assigned to the variable `images`. You need to pass them through `nyul_normalizer` to normalize them.

The code in the tutorial is an example of how to use the Python API of `intensity-normalization`, not an exact recipe for how to use it for your problem specifically. You need to try to understand the example conceptually, e.g., by reading the variable names, looking at the function calls and class instantiations, and looking at third-party documentation where necessary. Then you can adapt the example to your problem without my help.

If something is actively misleading or incorrect in the documentation or there is a bug with the software, then I can provide clarification or fix the bug; however, I don't have the time to teach general programming concepts.

Thank you for using `intensity-normalization`. Please raise an issue if there is a genuine bug or something incorrect in the documentation. Best of luck!

**Comment by hongshixia** [comment]
I'm sorry Sir, but I just understood what you said last time today. There is nothing wrong with your documentation and I have never questioned it. I'm sorry to have caused you a lot of trouble recently. I understood what you said last time today and made some changes, but my foundation was poor and he made an error. I hope you can give me some suggestions

![9938da74cf285303e602a5d1eba0578](https://user-images.githubusercontent.com/118377176/206512963-f473ff5c-6e2e-4406-894e-8af7a48575b4.jpg)
![9938da74cf285303e602a5d1eba0578](https://user-images.githubusercontent.com/118377176/206512963-f473ff5c-6e2e-4406-894e-8af7a48575b4.jpg)

**Reply by jcreinhold** [reply]
`nyul_norm` is a list not a function or class with a `__call__` method. Perhaps you want to use `nyul_normalizer` or `new_nyul_normalizer`.

Sorry. I can't respond to your questions anymore because they are about basic programming concepts and not `intensity-normalization` specifically. Best of luck.

**Comment by hongshixia** [comment]
Yes, Sir. Thank you

**Comment by hongshixia** [comment]
Hello, Sir. Thank you very much. I have gained a lot of knowledge from the previous communication with you. Now I've written the code and it's working fine on my computer. But I'm not sure if my code is correct. I'd like you to take a look at it for me. 
I have a question Sir, is the standard_histogram.npy of# 1 and# 2 marked in my code the same? 
At the end of the code run, all I have in my folder is standard_histogram.npy, I guess this file is saved at nyul_normalizer.save_standard_histogram("standard_histogram.npy"). I made a picture of it, and it looks like this, but I don't understand what it means. 
In my code I think the #2 and# 3 that I've labeled doesn't seem to work, right? Do you think so?

# load images

import nibabel as nib
image_paths = [r"D:/Desktop/STUDY/pythonProject4/learn/image1.nii",r"D:/Desktop/STUDY/pythonProject4/learn/image2.nii",r"D:/Desktop/STUDY/pythonProject4/learn/image3.nii",r"D:/Desktop/STUDY/pythonProject4/learn/image4.nii"]
images = [nib.load(image_path).get_fdata() for image_path in image_paths]

# normalize the images and save the standard histogram
from intensity_normalization.normalize.nyul import NyulNormalize
nyul_normalizer = NyulNormalize()
nyul_normalizer.fit(images)
normalized = [nyul_normalizer(image) for image in images]
nyul_normalizer.save_standard_histogram("standard_histogram.npy")# 1



# load new images (of the same modality) and normalize those
#I also put the images from the training set into the following sequence and then saved them together
new_image_paths = [  r"D:/Desktop/STUDY/pythonProject4/new/image1.nii",r"D:/Desktop/STUDY/pythonProject4/new/image2.nii",r"D:/Desktop/STUDY/pythonProject4/new/image3.nii",r"D:/Desktop/STUDY/pythonProject4/new/image4.nii",r"D:/Desktop/STUDY/pythonProject4/new/image5.nii"]
new_images = [nib.load(image_path).get_fdata() for image_path in new_image_paths]
normalized = [nyul_normalizer(image) for image in new_images]


new_nyul_normalizer = NyulNormalize()
new_nyul_normalizer.load_standard_histogram("standard_histogram.npy")# 2



a =[1,2,3,4,5]
for each in a:
    new_image = nib.load("image"+str(each)+".nii")
    b = 0
    normalize = normalized[b]
    nib.Nifti1Image(normalize, new_image.affine).to_filename("image"+str(each)+"normalized.nii")
    b = b+1
![image](https://user-images.githubusercontent.com/118377176/206858303-475d125c-0b13-4a3a-9a12-c824db6604c0.png)
![image](https://user-images.githubusercontent.com/118377176/206858412-eea33b43-39be-4e26-bcf9-de3c8e0d3936.png)
![image](https://user-images.githubusercontent.com/118377176/206858445-3185328e-47dd-47af-a6e3-78fd383c2f3f.png)
